### PR TITLE
Incorporate Scipy binning schemes into DdtHistLikelihood

### PR DESCRIPTION
**Pull request by jiwoncpark**
Currently, `DdtHistLikelihood` requires adjusting `nbins_hist`, with H0 inference results varying significantly depending on the choice. The numerical tuning can be tedious with hundreds of lens samples. Using the binning schemes such as Scott and Silverman can enable variable bin sizes and at least provide a starting point for tuning `nbins_hist`. Some side fixes:
- `bandwidth` attribute doesn't seem to be used
- KDE object's `logpdf` is more numerically stable than `log(pdf)`. 

Note I'm not dealing with `DdtHistKDELikelihood` at all, just `DdtHistLikelihood`. 

Please let me know if this sounds useful and I'll proceed to write the tests!

### CODE REVIEW

File: hierarc/Likelihood/LensLikelihood/ddt_hist_likelihood.py [hunk 1]
```diff
@@ -25,25 +26,36 @@ def __init__(self, z_lens, z_source, ddt_samples, kde_kernel=None, ddt_weights=N
...
+                                weights=ddt_weights)
+            vals = hist[0]
+            bins = [(h + hist[1][i+1])/2.0 for i, h in enumerate(hist[1][:-1])]
+            # ignore potential zero weights, sklearn does not like them
+            kde_bins = [b for v, b in zip(vals, bins) if v > 0]
+            kde_weights = [v for v in vals if v > 0]
+            self._kde = gaussian_kde(dataset=kde_bins, 
+                                     weights=kde_weights[:])
+        else:
+            self._kde = gaussian_kde(dataset=ddt_samples,
```
Comment (sibirrer) [inline hierarc/Likelihood/LensLikelihood/ddt_hist_likelihood.py:48]: can you add a test for this option please?

File: hierarc/Likelihood/LensLikelihood/ddt_hist_likelihood.py [hunk 2]
```diff
@@ -25,25 +26,36 @@ def __init__(self, z_lens, z_source, ddt_samples, kde_kernel=None, ddt_weights=N
         :param kde_kernel: string of KDE kernel type
         :param bandwidth: bandwith of kernel
         :param nbins_hist: number of bins in the histogram
-        :param normalized: bool, if True, returns the normalized likelihood, if False, separates the constant prefactor
-         (in case of a Gaussian 1/(sigma sqrt(2 pi)) ) to compute the reduced chi2 statistics
+        :param normalized: bool, if True, returns the normalized likelihood, 
+        if False, separates the constant prefactor (in case of a Gaussian
```
Comment (sibirrer) [inline hierarc/Likelihood/LensLikelihood/ddt_hist_likelihood.py:30]: formatting: If the documentation goes to the second line, please insert by one spacing to be conformal with the sphinx documentation scheme

**Review by sibirrer — CHANGES_REQUESTED** [review]
This sounds useful indeed and I can approve it if you add a test function for the new feature. I also wonder what the difference in speed is as this might become a bit an issue for a lot of lenses and hierarchical sampling. Thank you very much!

**Comment by sibirrer** [comment]
ah, and I see the tests fail: https://travis-ci.com/github/sibirrer/hierArc/builds/210324201 the keyword argument 'bandwidth' is not required so you can remove it from being passed to the call in the tests.

File: hierarc/Likelihood/LensLikelihood/ddt_hist_likelihood.py (hunk 2 above)
Comment (jiwoncpark) [reply hierarc/Likelihood/LensLikelihood/ddt_hist_likelihood.py]: This only applies for parameter docstrings (that start with colons), correct? If so, done!

File: hierarc/Likelihood/LensLikelihood/ddt_hist_likelihood.py (hunk 1 above)
Comment (jiwoncpark) [reply hierarc/Likelihood/LensLikelihood/ddt_hist_likelihood.py:52]: Tests added! I used unittest but pytest should still recognize it.

**Comment by sibirrer** [comment]
tests failed. You should be able to execute the tests locally before pushing

**Comment by jiwoncpark** [comment]
> tests failed. You should be able to execute the tests locally before pushing

My bad, I only ran the test module I fixed locally. Now all tests are passing locally.

**Review by sibirrer — APPROVED** [review]

**Comment by sibirrer** [comment]
Nice testing format @jiwoncpark ! Thank you very much!

**Pull request merged**